sns/parity_math.c

```diff
@@ -276,15 +526,70 @@ static int vandmat_norm(struct m0_matrix *m)
 				m0_matrix_cols_operate(m, x, y, gsub, 0, gmul,
                                             *m0_matrix_elem_get(m, x, y), gsub);
 
-		/* Assert if units are not configured properly. */
-		M0_ASSERT(check_row_is_id(m, y));
+		/* Check if units are configured properly. */
+		if (check_row_is_id(m, y) == false)
+			return M0_ERR_INFO(-EINVAL, "Units are not configured "
+					   "properly. row=%u of matrix=%p is "
+					   "not part of identity matrix",
+					   y, m);
 	}
 
 	return 0;
 }
+#endif /* !ISAL_ENCODE_ENABLED */
+
+#if ISAL_ENCODE_ENABLED
+static bool parity_math_invariant(const struct m0_parity_math *math)
+{
+	return  _0C(math != NULL) && _0C(math->pmi_data_count >= 1) &&
+		_0C(math->pmi_parity_count >= 1) &&
+		_0C(math->pmi_data_count >= math->pmi_parity_count) &&
+		_0C(math->pmi_data_count <= SNS_PARITY_MATH_DATA_BLOCKS_MAX);
+}
+
+static int isal_encode_data_update(struct m0_buf *dest_bufs, struct m0_buf *src_buf,
+				   uint32_t vec_idx, uint8_t *g_tbls,
+				   uint32_t data_nr, uint32_t dest_nr)
+{
+	uint32_t i;
+	uint32_t block_size;
+
+	M0_ENTRY("dest_bufs=%p, src_buf=%p, vec_idx=%u, "
+		 "g_tbls=%p, data_nr=%u, dest_nr=%u",
+		 dest_bufs, src_buf, vec_idx, g_tbls, data_nr, dest_nr);
+
+	M0_PRE(dest_bufs != NULL);
+	M0_PRE(src_buf != NULL);
+	M0_PRE(g_tbls != NULL);
+
+	uint8_t	 *dest_frags[dest_nr];
+
+	block_size = (uint32_t)src_buf->b_nob;
+
+	for (i = 0; i < dest_nr; ++i) {
+		if (block_size != dest_bufs[i].b_nob)
+			return BLOCK_SIZE_ERR_INFO("dest_bufs", block_size, i,
+						   dest_bufs[i].b_nob);
```

Comment:
This is a logical error. It should trigger an assertion. Similarly in other places.

Reply:
This change was done as part of PR #741.

Reply:
Done